### ISSUE TITLE
BERT Models: Huge difference in last hidden states of similar examples

### ISSUE THREAD

**Issue by vakkov** [OpenNMT/CTranslate2#1296]
Hi!
First and foremost, thanks for the great package!

We've encountered a peculiar problem while running BERT and similar transformers through CTranslate2; As you will find out in the code below, running almost identical batches (2 out of 3 inputs are identical) produces very different results; something that does not happen when inferencing the HuggingFace BERT from the documentation/ tutorial. I would expect that the hidden states of the first two samples would be quite close/ similar (as is the case with HF). 

```
import ctranslate2
import numpy as np
import torch
import transformers

device = "cpu"

def read_array(output):
    if device == "cuda":
        return torch.as_tensor(output, device=device)
    else:
        return np.array(output)

encoder = ctranslate2.Encoder("bert-base-uncased-yelp-polarity", device=device)

model_name = "textattack/bert-base-uncased-yelp-polarity"
hf_model = transformers.AutoModelForSequenceClassification.from_pretrained(model_name)
hf_model.eval()
tokenizer = transformers.AutoTokenizer.from_pretrained(model_name)

inputs = ["How are you?", "This sucks.", "I'm so happy for you!"]
inputs2 = ["How are you?", "This sucks.", "Why though?"]

tokens = tokenizer(inputs, return_tensors="np", padding=True, truncation=True, max_length=512)
token2 = tokenizer(inputs2, return_tensors="np", padding=True, truncation=True, max_length=512)

hf_tokens = tokenizer(inputs, return_tensors="pt", padding=True, truncation=True, max_length=512)
hf_tokens2 = tokenizer(inputs2, return_tensors="pt", padding=True, truncation=True, max_length=512)

output = read_array(encoder.forward_batch(tokens['input_ids']).last_hidden_state)
output2 = read_array(encoder.forward_batch(token2['input_ids']).last_hidden_state)

with torch.no_grad():
    hf_output = hf_model.bert.forward(**hf_tokens)
    hf_output2 = hf_model.bert.forward(**hf_tokens2)

#FILL output2 (3, 6, 768) to match the shape of output (3, 10, 768)
output2 = np.concatenate((output2, np.zeros((3, 4, 768))), axis=1)
#output2 = torch.as_tensor(output2, device=device)
diff_ctranslate = np.abs(output - output2)

hf_output2 = hf_output2.last_hidden_state.detach().numpy()
#FILL hf_output2 (3, 6, 768) to match the shape of hf_output (3, 10, 768)
hf_output2 = np.concatenate((hf_output2, np.zeros((3, 4, 768))), axis=1)
#hf_output2 = torch.as_tensor(hf_output2, device=device)
diff_hf = np.abs(hf_output.last_hidden_state.detach().numpy() - hf_output2)

import pdb; pdb.set_trace()
```

```
(Pdb) diff_hf
array([[[2.98023224e-08, 5.96046448e-08, 5.96046448e-08, ...,
         1.11758709e-08, 7.45058060e-09, 2.98023224e-08],
        [6.70552254e-08, 1.04308128e-07, 1.41561031e-07, ...,
         4.09781933e-08, 1.35041773e-08, 8.94069672e-08],
        [5.58793545e-08, 1.49011612e-08, 4.47034836e-08, ...,
         7.45058060e-09, 3.72529030e-08, 0.00000000e+00],
        ...,
        [8.38382449e-03, 3.04741561e-01, 1.24678187e-01, ...,
         3.12836654e-02, 1.87378544e-02, 1.46343589e-01],
        [2.68252857e-04, 3.32778931e-01, 1.39690027e-01, ...,
         3.26546878e-02, 3.70148011e-02, 1.51049569e-01],
        [2.83848960e-02, 3.11481982e-01, 1.67232439e-01, ...,
         4.96172495e-02, 1.41829997e-02, 1.34320632e-01]],

       [[5.96046448e-08, 1.19209290e-07, 5.96046448e-08, ...,
         0.00000000e+00, 0.00000000e+00, 1.19209290e-07],
        [5.96046448e-08, 0.00000000e+00, 1.78813934e-07, ...,
         3.91155481e-08, 5.96046448e-08, 5.96046448e-08],
        [2.68220901e-07, 0.00000000e+00, 1.78813934e-07, ...,
         1.11758709e-07, 5.96046448e-08, 5.96046448e-08],
        ...,
        [3.96514803e-01, 1.20959663e+00, 5.56879461e-01, ...,
         5.74268252e-02, 5.17984629e-01, 7.45872438e-01],
        [3.94066989e-01, 1.20485914e+00, 5.54322422e-01, ...,
         5.23871034e-02, 5.18572628e-01, 7.43265748e-01],
        [3.94618690e-01, 1.20497525e+00, 5.54426849e-01, ...,
         5.29986843e-02, 5.18883288e-01, 7.42800772e-01]],

       [[1.42525335e-01, 1.20464973e+00, 3.03964064e-01, ...,
         1.38388488e-01, 4.77911960e-01, 6.56643605e-01],
        [1.83270124e-01, 8.27781398e-01, 1.97174547e-01, ...,
         1.97876673e-01, 3.15828778e-01, 7.11090401e-01],
        [2.49992363e-01, 9.17543977e-01, 2.34332152e-01, ...,
         1.86537050e-01, 3.01875435e-01, 7.36145050e-01],
        ...,
        [2.38040492e-01, 9.33259666e-01, 1.63673088e-01, ...,
         2.53996193e-01, 4.01549369e-01, 5.11862576e-01],
        [9.69869941e-02, 8.73746037e-01, 1.13427311e-01, ...,
         1.52672216e-01, 5.15702188e-01, 6.42594337e-01],
        [4.08798419e-02, 8.73961568e-01, 1.45702690e-01, ...,
         1.70884669e-01, 1.57639399e-01, 3.43730628e-01]]])
```
         
         
         
```
(Pdb) diff_ctranslate
array([[[0.02916592, 0.36293509, 0.12465167, ..., 0.03291841,
         0.10621519, 0.14973479],
        [0.0828281 , 0.29909007, 0.10487518, ..., 0.01129711,
         0.12281137, 0.15931579],
        [0.10475164, 0.08716819, 0.04082789, ..., 0.02784223,
         0.11741078, 0.17508738],
        ...,
        [0.04118664, 0.01387221, 0.01139041, ..., 0.04329493,
         0.13025621, 0.0041482 ],
        [0.05121706, 0.03593045, 0.01179469, ..., 0.052505  ,
         0.12422864, 0.00272606],
        [0.02618104, 0.02223308, 0.01202715, ..., 0.03945384,
         0.12365534, 0.01045229]],

       [[0.27138749, 0.20181757, 0.1586653 , ..., 0.47498369,
         0.28579104, 0.03163797],
        [0.27205806, 0.22677261, 0.16803446, ..., 0.49863124,
         0.26618633, 0.00459033],
        [0.31137139, 0.02918708, 0.01648539, ..., 0.39186636,
         0.20487279, 0.00464034],
        ...,
        [0.13749281, 0.85713243, 0.52486527, ..., 0.43861988,
         0.3147983 , 0.26252115],
        [0.11911354, 0.74020255, 0.55119479, ..., 0.4111037 ,
         0.36102304, 0.19190362],
        [0.20134039, 0.56342733, 0.51808345, ..., 0.29945564,
         0.41682807, 0.08429318]],

       [[0.14305582, 1.12097611, 0.27286507, ..., 0.10952042,
         0.49072006, 0.63290307],
        [0.15372671, 0.80011579, 0.19787918, ..., 0.18482324,
         0.31356857, 0.68834567],
        [0.22959686, 0.87788446, 0.22443378, ..., 0.16879096,
         0.29412676, 0.71146711],
        ...,
        [0.23804054, 0.93325943, 0.16367307, ..., 0.2539959 ,
         0.40154937, 0.51186246],
        [0.09698693, 0.8737458 , 0.11342725, ..., 0.15267208,
         0.51570213, 0.64259404],
        [0.04087965, 0.87396133, 0.14570253, ..., 0.17088449,
         0.15763937, 0.34373054]]])
```

**Comment by guillaumekln** [comment]
Hi,

It's the same issue as #1286.

The input ids contain padding tokens, but you are not setting the `lengths` argument in `forward_batch` to mask these tokens.

The easiest solution is to tokenize the inputs into Python lists and let CTranslate2 do the padding:

```diff
24,25c24,25
< tokens = tokenizer(inputs, return_tensors="np", padding=True, truncation=True, max_length=512)
< token2 = tokenizer(inputs2, return_tensors="np", padding=True, truncation=True, max_length=512)
---
> tokens = tokenizer(inputs, truncation=True, max_length=512)
> token2 = tokenizer(inputs2, truncation=True, max_length=512)
```

If you want to keep adding the padding tokens in your code, you can build the lengths argument like this:

```diff
24,26d23
< tokens = tokenizer(inputs, return_tensors="np", padding=True, truncation=True, max_length=512)
< token2 = tokenizer(inputs2, return_tensors="np", padding=True, truncation=True, max_length=512)
< 
30,31c27,34
< output = read_array(encoder.forward_batch(tokens['input_ids']).last_hidden_state)
< output2 = read_array(encoder.forward_batch(token2['input_ids']).last_hidden_state)
---
> input_ids = ctranslate2.StorageView.from_array(hf_tokens["input_ids"].to(torch.int32).numpy())
> input_ids2 = ctranslate2.StorageView.from_array(hf_tokens2["input_ids"].to(torch.int32).numpy())
> 
> lengths = ctranslate2.StorageView.from_array(hf_tokens["attention_mask"].sum(1, dtype=torch.int32).numpy())
> lengths2 = ctranslate2.StorageView.from_array(hf_tokens2["attention_mask"].sum(1, dtype=torch.int32).numpy())
> 
> output = read_array(encoder.forward_batch(input_ids, lengths).last_hidden_state)
> output2 = read_array(encoder.forward_batch(input_ids2, lengths2).last_hidden_state)
```

**Comment by vakkov** [comment]
Thanks for the help and sorry for the issue duplication!

I'd like to use the opportunity to ask you if there are any ways to add a classification head (such as the one below) as part of the converted (to ctranslate) model since throwing tensors at a HF classifier is sort of slow...

```
  (classifier): XLMRobertaClassificationHead(
    (dense): Linear(in_features=768, out_features=768, bias=True)
    (dropout): Dropout(p=0.1, inplace=False)
    (out_proj): Linear(in_features=768, out_features=3, bias=True)
  )
```

Should we try adding it as a pooler when converting the current model or should I write a separate classifier.cc with forward inference methods and so on.

P.S. Another issue we are facing is speed-related - the Ctranslate2 model seems to be on par with (and sometimes even slower than) our HuggingFace transformer when ran on a relatively big dataset with a batch size of 12 (and no token padding, ofc :P)

**Comment by guillaumekln** [comment]
> since throwing tensors at a HF classifier is sort of slow

Is it? The classification head is small compared to the rest of the model so I don't think it makes a big impact. It is not worth adding to CTranslate2 just for the performance in my opinion.

If you just want to get rid of PyTorch, maybe you can load and run the weights with Numpy?

> P.S. Another issue we are facing is speed-related - the Ctranslate2 model seems to be on par with (and sometimes even slower than) our HuggingFace transformer when ran on a relatively big dataset with a batch size of 12 (and no token padding, ofc :P)

Can you share the code you are running for the comparison?

**Comment by guillaumekln** [comment]
I'm closing this issue since the difference in the hidden states was explained and addressed. Feel free to open another issue if there are performance issues.

**Comment by poornaprudhvigu-kore** [comment]
Hey @guillaumekln, 

Thanks for the detailed response for the issue. With your approach we are missing out on batching the sentences to be encoded right ? We have to all the time seperately encode them, Can you suggest a way out here ?
